DOC-43/Pregel algorithms - restructured content (#1047)
arangodb/docs: 3.9/graphs-pregel-algorithms.md

```js
var pregel = require("@arangodb/pregel");

// connected components
pregel.start("connectedcomponents", "graphname", {resultField: "component"});

// weakly connected components
pregel.start("wcc", "graphname", {resultField: "component_weak"});

// strongly connected components
pregel.start("scc", "graphname", {resultField: "component_strong"});
```

### Hyperlink-Induced Topic Search (HITS)

HITS is a link analysis algorithm that rates Web pages, developed by
Jon Kleinberg. The algorithm is also known as _Hubs and Authorities_.

The idea behind Hubs and Authorities comes from the typical structure of the web:
Certain websites known as hubs, serve as large directories that are not actually
authoritative on the information that they hold. These hubs are used as
compilations of a broad catalog of information that leads users direct to other
authoritative webpages.

The algorithm assigns each vertex two scores: The authority score and the

> hub score. The authority score rates how many good hubs point to a particular
> vertex (or webpage), the hub score rates how good (authoritative) the vertices
> pointed to are. Also see

Reviewer: ```suggestion
hub score. The authority score of a vertex rates the total hub score of vertices pointing to that vertex; the hub score rates the total authority score of vertices pointed by it. Also see
```

[en.wikipedia.org/wiki/HITS_algorithm](https://en.wikipedia.org/wiki/HITS_algorithm){:target="_blank"}

ArangoDB's version of the algorithm converges after a certain amount of time.
The parameter *threshold* can be used to set a limit for the convergence
(measured as maximum absolute difference of the hub and authority scores
between the current and last iteration).

When you specify the result field name, the hub score will be stored in
`<resultField>_hub` and the authority score in `<resultField>_auth`.

The algorithm can be executed like this:

```js
var pregel = require("@arangodb/pregel");

> var handle = pregel.start("hits", "yourgraph", {threshold:0.00001, resultField: "score"});

Reviewer: for consistency: 
```suggestion
var handle = pregel.start("hits", "graphname", {threshold:0.00001, resultField: "score"});
```

(or closeness). The closeness of a vertex in a graph is the inverse average
length of the shortest path between the vertex and all other vertices.
For vertices *x*, *y* and shortest distance `d(y, x)` it is defined as:

![Vertex Closeness Formula](images/closeness.png)

Effective Closeness approximates the closeness measure. The algorithm works by
iteratively estimating the number of shortest paths passing through each vertex.
The score will approximates the real closeness score, since it is not possible
to actually count all shortest paths due to the horrendous `O(n^2 * d)` memory
requirements. The algorithm is from the paper
*Centralities in Large Networks: Algorithms and Observations (U Kang et.al. 2011)*.

ArangoDBs implementation approximates the number of shortest path in each
iteration by using a HyperLogLog counter with 64 buckets. This should work well
on large graphs and on smaller ones as well. The memory requirements should be
**O(n * d)** where *n* is the number of vertices and *d* the diameter of your
graph. Each vertex will store a counter for each iteration of the algorithm.

The algorithm can be used like this:

```js
const pregel = require("@arangodb/pregel");

> const handle = pregel.start("effectivecloseness", "yourgraph", {resultField: "closeness"});

Reviewer: ```suggestion
const handle = pregel.start("effectivecloseness", "graphname", {resultField: "closeness"});
```

It measures the number of times a vertex is part of shortest paths between any
pairs of vertices. For a vertex *v* betweenness is defined as:

![Vertex Betweenness Formula](images/betweenness.png)

Where the &sigma; represents the number of shortest paths between *x* and *y*,
and &sigma;(v) represents the number of paths also passing through a vertex *v*.
By intuition a vertex with higher betweenness centrality will have more
information passing through it.

**LineRank** approximates the random walk betweenness of every vertex in a
graph. This is the probability that someone starting on an arbitrary vertex,
will visit this node when he randomly chooses edges to visit.

The algorithm essentially builds a line graph out of your graph
(switches the vertices and edges), and then computes a score similar to PageRank.
This can be considered a scalable equivalent to vertex betweenness, which can
be executed distributedly in ArangoDB. The algorithm is from the paper
*Centralities in Large Networks: Algorithms and Observations (U Kang et.al. 2011)*.

```js
const pregel = require("@arangodb/pregel");

> const handle = pregel.start("linerank", "yourgraph", {resultField: "linerank"});

Reviewer: ```suggestion
const handle = pregel.start("linerank", "graphname", {resultField: "linerank"});
```

If a vertex keeps its ID 20 times or more in a row, it does not send its ID. Vertices that did not obtain any IDs do not update their ID and do not send it.

The algorithm runs until it converges, which likely never really happens on
large graphs. Therefore you need to specify a maximum iteration bound.
The default bound is 500 iterations, which is too large for
common applications. 

The algorithm should work best on undirected graphs. On directed
graphs, the resulting partition into communities might change, if the number 
of performed steps changes. How strong the dependence is
may be influenced by the density of the graph.

```js
const pregel = require("@arangodb/pregel");

> const handle = pregel.start("labelpropagation", "yourgraph", {maxGSS: 100, resultField: "community"});

Reviewer: ```suggestion
const handle = pregel.start("labelpropagation", "graphname", {maxGSS: 100, resultField: "community"});
```

```

#### Speaker-Listener Label Propagation

The [Speaker-listener Label Propagation](https://arxiv.org/pdf/1109.5720.pdf){:target="_blank"}
(SLPA) can be used to implement community detection. It works similar to the
label propagation algorithm, but now every node additionally accumulates a
memory of observed labels (instead of forgetting all but one label).

Before the algorithm run, every vertex is initialized with an unique ID
(the initial community label).
During the run three steps are executed for each vertex:

1. Current vertex is the listener all other vertices are speakers
2. Each speaker sends out a label from memory, we send out a random label with a
   probability proportional to the number of times the vertex observed the label
3. The listener remembers one of the labels, we always choose the most
   frequently observed label

```js
const pregel = require("@arangodb/pregel");

> const handle = pregel.start("slpa", "yourgraph", {maxGSS:100, resultField: "community"});

Reviewer: ```suggestion
const handle = pregel.start("slpa", "graphname", {maxGSS:100, resultField: "community"});
```

```

You can also execute SLPA with the `maxCommunities` parameter to limit the
number of output communities. Internally the algorithm will still keep the
memory of all labels, but the output is reduced to just he `n` most frequently
observed labels.

```js
const pregel = require("@arangodb/pregel");

> const handle = pregel.start("slpa", "yourgraph", {maxGSS: 100, resultField: "community", maxCommunities: 1});

Reviewer: ```suggestion
const handle = pregel.start("slpa", "graphname", {maxGSS: 100, resultField: "community", maxCommunities: 1});
```

The algorithm stops when at least one of the two conditions is satisfied:
- The maximum number of iterations is reached. This is the same parameter `maxGSS` as for the other algorithms.
- Every vertex changes its rank in the last iteration by less than a certain threshold. The default threshold is  0.00001, a custom value can be set with the parameter `threshold`.
```js
var pregel = require("@arangodb/pregel");
pregel.start("pagerank", "graphname", {maxGSS: 100, threshold: 0.00000001, resultField: "rank"})
```

#### Seeded PageRank

It is possible to specify an initial distribution for the vertex documents in
your graph. To define these seed ranks / centralities you can specify a
`sourceField` in the properties for this algorithm. If the specified field is
set on a document _and_ the value is numeric, then it will be used instead of
the default initial rank of `1 / numVertices`.

```js
var pregel = require("@arangodb/pregel");
pregel.start("pagerank", "graphname", {maxGSS: 20, threshold: 0.00000001, sourceField: "seed", resultField: "rank"})
```

### Single-Source Shortest Path

> Calculates the shortest path length between the given source and all other vertices, called _targets_. The result is written to the specified property of the respective target.

Reviewer: ```suggestion
Calculates the lengths of shortest paths from the given source to all other vertices, called _targets_. The result is written to the specified property of the respective target.
```

The distance to the source vertex itself is returned as `0` and a length above
`9007199254740991` (max safe integer) means that there is no path from the source to the vertex in the graph.
a pair of vertices.

The algorithm runs until all distances are computed. The number of iterations is bounded by the

> diameter (the longest shortest path) of your graph.

Reviewer: ```suggestion
diameter of your graph (the longest distance between two vertices).
```

be executed distributedly in ArangoDB. The algorithm is from the paper
*Centralities in Large Networks: Algorithms and Observations (U Kang et.al. 2011)*.

```js
const pregel = require("@arangodb/pregel");
const handle = pregel.start("linerank", "yourgraph", {resultField: "linerank"});
```

### Community Detection

Graphs based on real world networks often have a community structure.
This means it is possible to find groups of vertices such that each vertex
group is internally more densely connected than outside the group.
This has many applications when you want to analyze your networks, for example
Social networks include community groups (the origin of the term, in fact)
based on common location, interests, occupation, etc.

#### Label Propagation

*Label Propagation* can be used to implement community detection on large
graphs. The algorithm assigns a community, more precisely, a Community ID 
(a natural number), to every vertex in the graph. 
The idea is that each vertex should be in the community that most of
its neighbors are in. 

At first, the algorithm assigns unique initial Community IDs to the vertices. 
There is no guarantee that a vertex obtains the same initial 

> ID in two different runs of the algorithm, even if the graph does not change
> (although, it may often happen). Moreover, there is no guarantee on a particular

Reviewer: ```suggestion
ID in two different runs of the algorithm, even if the graph does not change 
(because the sharding may change). Moreover, there is no guarantee on a particular
```

distribution of the initial IDs over the vertices.

Then, in each iteration, a vertex sends its current Community
ID to all its neighbor vertices. After that each vertex adopts the Community ID it
received most frequently in the last step. 

> The details are somewhat subtle. If a vertex obtains only one ID and the ID of the vertex from the previous step, its old ID, is less than the obtained ID, the old ID is kept. (IDs are numbers and thus comparable to each other.) If a vertex obtains more than one ID, its new ID is the lowest ID among the most frequently obtained IDs. (For example, if the obtained IDs are 1, 2, 2, 3, 3, then 2 is the new ID. ) If, however, no ID arrives more than once, the new ID is the minimum of the lowest obtained IDs and the old ID. (For example, if the old ID is 5 and the obtained IDs are 3, 4, 6, then the new ID is 3. If the old ID is 2, it is kept.) 

Reviewer: ```suggestion
Note that, in a usual implementation of Label Propagation, if there are multiple most frequently received Community IDs, one is chosen randomly. An advantage of our implementation is that this choice is deterministic. This comes for the price that the choice rules are somewhat involved: If a vertex obtains only one ID and the ID of the vertex from the previous step, its old ID, is less than the obtained ID, the old ID is kept. (IDs are numbers and thus comparable to each other.) If a vertex obtains more than one ID, its new ID is the lowest ID among the most frequently obtained IDs. (For example, if the obtained IDs are 1, 2, 2, 3, 3, then 2 is the new ID. ) If, however, no ID arrives more than once, the new ID is the minimum of the lowest obtained IDs and the old ID. (For example, if the old ID is 5 and the obtained IDs are 3, 4, 6, then the new ID is 3. If the old ID is 2, it is kept.) 
```

---
layout: default
description: Pregel algorithms can be used for graph exploration, path finding, analytics queries, and much more. 
---
# Pregel Algorithms

{{ page.description }}
{:class="lead"}

> Pregel algorithms are used in scenarios where you need to do an

Reviewer: The term "Pregel" is used in our docs both in the sense of a computation model and of our implementation of this model. For me, it's not always clear what meaning is being used. Because we cannot change the well-known term "Pregel" as the computation model, I suggest to use another term at least in the documentation, e.g., "Arango-Pregel" (this should probably be discussed with our management).

This means it is possible to find groups of vertices such that each vertex
group is internally more densely connected than outside the group.
This has many applications when you want to analyze your networks, for example
Social networks include community groups (the origin of the term, in fact)
based on common location, interests, occupation, etc.

#### Label Propagation

*Label Propagation* can be used to implement community detection on large
graphs. The algorithm assigns a community, more precisely, a Community ID 
(a natural number), to every vertex in the graph. 
The idea is that each vertex should be in the community that most of
its neighbors are in. 

At first, the algorithm assigns unique initial Community IDs to the vertices. 
There is no guarantee that a vertex obtains the same initial 
ID in two different runs of the algorithm, even if the graph does not change
(although, it may often happen). Moreover, there is no guarantee on a particular
distribution of the initial IDs over the vertices.

Then, in each iteration, a vertex sends its current Community
ID to all its neighbor vertices. After that each vertex adopts the Community ID it

> received most frequently in the last step. If a vertex obtains more than one
> most frequent IDs, it chooses the lowest number (as IDs are numbers). If no ID arrived more 
> than once and the ID of the vertex from the previous step is less than the
> lowest obtained ID number, the old ID is kept. 

Reviewer: ```suggestion
received most frequently in the last step. If a vertex obtains more than one
IDs, its new ID becomes the lowest obtained ID (IDs are numbers and thus comparable to each other). If only one ID arrived and the ID of the vertex from the previous step is less than the
obtained ID, the old ID is kept. If a vertex keeps its ID 20 times or more in a row, it does not send its ID. Vertices that did not obtain any IDs do not update their ID and do not send it.
```
